\documentclass[12pt,Bold,letterpaper,TexShade]{mcgilletdclass}

\usepackage{amsfonts}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{caption}
\usepackage{color}
\usepackage{enumitem}
\usepackage{epsfig}
\usepackage{framed}
\usepackage{geometry}
\usepackage{graphicx}
\usepackage{listings}
\usepackage{subcaption}

\input{def_mcgill}

\addtolength{\hoffset}{0pt}		% Have you configured your TeX system for proper page alignment? See the McGillETD documentation
\addtolength{\voffset}{0pt}

%-----------------------------------------------------------------------------
% Student info
%-----------------------------------------------------------------------------

\SetTitle{\huge{Computational challenges in genome wide association studies: data processing, variant annotation and epistasis}}
\SetAuthor{Pablo Cingolani}
\SetDegreeType{PhD.}
\SetDepartment{School of Computer Science}
\SetUniversity{McGill University}
\SetUniversityAddr{Montreal,Quebec}
\SetThesisDate{March 2015}
\SetRequirements{A thesis submitted to McGill University in partial fulfillment of the requirements of the degree of Doctor of Philosophy}
\SetCopyright{Pablo Cingolani 2015}

\makeindex[keylist]
\makeindex[abbr]
\listfiles

%-----------------------------------------------------------------------------
% Code linsting options
%-----------------------------------------------------------------------------

\definecolor{dkgreen}{rgb}{0,0.6,0}
\definecolor{gray}{rgb}{0.5,0.5,0.5}
\definecolor{mauve}{rgb}{0.58,0,0.82}

\lstset{frame=tb,
  language=Java,
  aboveskip=3mm,
  belowskip=3mm,
  showstringspaces=false,
  columns=flexible,
  basicstyle={\small\linespread{1.0}\ttfamily,},
  numbers=none,
  numberstyle=\tiny\color{gray},
  keywordstyle=\color{blue},
  commentstyle=\color{dkgreen},
  stringstyle=\color{mauve},
  breaklines=true,
  breakatwhitespace=true,
  numbers=left,
  stepnumber=1,
  firstnumber=1,
  numberfirstline=true
  tabsize=3
}

\renewcommand\lstlistingname{Listing}
\renewcommand\lstlistlistingname{Listings}

%-----------------------------------------------------------------------------
% Document stats here
%-----------------------------------------------------------------------------

\begin{document}

\maketitle

%-----------------------------------------------------------------------------
% Input any special commands below
%-----------------------------------------------------------------------------

% Conditional expression for faster build during "development" cycle
% See http://tex.stackexchange.com/questions/5894/latex-conditional-expression
\newif\ifthesis
%\thesistrue
\thesisfalse

% Final version?
\newif\iffinal
%\finaltrue
\finalfalse

\ifthesis
	% Only show these sections when we are building the 'thesis' version

	\begin{romanPagenumber}{2}
	
	\iffinal

		%-----------------------------------------------------------------------------
		% Acknowledgements
		%-----------------------------------------------------------------------------
		
		\SetAcknowledgeName{\MakeUppercase{Acknowledgements}}
		\SetAcknowledgeText{
		}
		\Acknowledge
	
	\fi

	%-----------------------------------------------------------------------------
	%         English Abstract
	%-----------------------------------------------------------------------------
	
	\SetAbstractEnName{\MakeUppercase{Abstract}}%
	\SetAbstractEnText{ 
Abundant genome sequence information from large cohorts of individuals can now be routinely obtained and this information is poised to ease the identification of genetic variations linked to complex disease. 
In this work, I investigate the computational and statistical challenges involved in the analysis of large genomic datasets and I tackle three different aspects of the analysis, each of them having very different characteristics.
First, in order to analyse large amounts of data from genomic studies we design a programming language, BigDataScript, that simplifies the creation of robust and scalable data analysis pipelines.
Second, we create genomic variant annotation and prioritization methods (SnpEff and SnpSift) that help to calculate putative genetic effects and estimate the genetic impact of variants.
Finally, we address the problem of finding associations between interacting genetic loci and disease by proposing a methodology that combines population-level genetic information with evolutionary information in order to increase the statistical power in epistatic genome wide association studies 
	}
	\AbstractEn%
	
	%-----------------------------------------------------------------------------
	%         French Abstract
	%-----------------------------------------------------------------------------
	
	\SetAbstractFrName{\MakeUppercase{ABR\'{E}G\'{E}}}%
	\SetAbstractFrText{ 
	}
	\AbstractFr%
	
	%-----------------------------------------------------------------------------
	% Tables
	%-----------------------------------------------------------------------------
	
	\TOCHeading{\MakeUppercase{Table of Contents}}
	% \LOTHeading{\MakeUppercase{List of Tables}}
	\LOFHeading{\MakeUppercase{List of Figures and Tables}}
	\tableofcontents 
	%\listoftables 
	\listoffigures
	
	\end{romanPagenumber}

\else
	% Skip all the previous sections
\fi

%-----------------------------------------------------------------------------
% Chapters
%-----------------------------------------------------------------------------

\ifthesis
	% Show all chapters on 'thesis' mode
	\include{chapters/intro}
	\include{chapters/intro_variants}
	\include{chapters/intro_epistasis}
	\include{chapters/intro_coevolution}
	\include{chapters/intro_roadmap}
	\include{chapters/bds}
	\include{chapters/snpeff}
	\include{chapters/gwas}
	\include{chapters/gwas_supp}
	\include{chapters/conclusions}
\else
	\include{chapters/intro}
	\include{chapters/intro_variants}
	% Only show the chapter we are currently working on
\fi

\ifthesis
	% Show all appendices & bibliography in 'thesis' mode

	%-----------------------------------------------------------------------------
	% Appendices
	%-----------------------------------------------------------------------------
	%\ETDAppendix{Appendix A}{ \include{appendix/algorithm_details} }%
\else
	% Skip all appendices and bibliography
\fi
	
%-----------------------------------------------------------------------------
% Bibliography. You need to run the following 
% command: 
%       bibtex thesis_mcgill
%-----------------------------------------------------------------------------
\bibHeading{References}
\bibliographystyle{plain}
\bibliography{mcgilletd}		% File: mcgilletd.bib

\end{document}
