%---
\section{Epistasis \label{sec:epi}}
%---

At the beginning of the 20th century some deviations form classical Mendelian inheritance were characterized.
William Bateson first described epistasis in 1907 \cite{tyler2009shadows} assessing a discrepancy between the prediction of segregation ratios assuming individual genes and the real outcome \cite{phillips2008epistasis}.
The term epistasis, which literally means ``standing upon", was used to describe ``characters'' layered on top of other each other thus masking their expression.
This original definition describing the situation in which the actions of one locus mask the allelic effects of another locus is an extension of dominance where a completely dominant alleles mask the effects of the recessive allele at the same locus \cite{carlborg2004epistasis, cordell2002epistasis}.

Nowadays the term epistasis is not only used to describe the original definition \cite{cordell2002epistasis}, but also often interpreted as mutations in two genes producing a phenotype that is surprising considering the individual effect of each mutation.
Furthermore, in some contexts epistasis is used to refer to a broad range gene-gene interactions, many complex interactions among genetic loci or even interaction between gene and the environment \cite{phillips2008epistasis}.
Three categories of epistasis commonly used by geneticists are \cite{phillips2008epistasis, zhao2006test}: 

\begin{itemize}
	\item Functional epistasis: The molecular interactions between proteins, usually consisting of proteins within the same pathway or of within a complex.
	\item Compositional epistasis: Describes the traditional usage of epistasis as described by Bateson (i.e. masking of one allelic effect by an allele at another locus).
	\item Statistical epistasis: This terminology is attributed to Fisher and defined as a deviance from genetic additive effects.
\end{itemize}

These concepts imply that analysis of epistasis can be used to infer functional relationships between genes \cite{mani2008defining}, genetic pathways' structure and function as well as evolutionary dynamics \cite{phillips2008epistasis}. 
Some authors even relate the analysis of epistatic gene interactions patterns to the fundamentals of systems biology \cite{phillips2008epistasis}.

Epistasis can be classified by the way a deviation of a double-mutant (having one mutation at each loci) organism's phenotype differed from the expected neutral (non-interacting) phenotype \cite{mani2008defining}.
An interaction is known as ``synergistic" or ``synthetic" when the double mutant has a more extreme phenotype than expected based on the phenotype of the two individual mutants.
When the phenotype is less severe than expected, then there is a ``alleviating" or ``diminishing returns" interaction, which is often attributed to gene products operating in series within the pathway. 
A typical example is a mutation in one gene impairing a whole pathway, thus masking the consequence of mutations in other genes of the same pathway. \cite{mani2008defining}.

Often, a phenotype in human genetics is qualitative and dichotomous, for instance indicating presence or absence of disease. \cite{cordell2002epistasis}.
Thus mathematical models calculating the joint action of more than one loci focus on the penetrance (the probability of developing disease given genotype).
Assuming an allele is required at both loci in order to express the trait, the effect of allele $A$ can only be observed when allele $B$ is also present.
This means that the effect at locus $A$ appears masked by locus $B$ and vice-versa \cite{cordell2002epistasis}, which is not precisely analogous to
what Bateson described since in Bateson's definition if factor $B$ is epistatic to factor $A$, then factor $A$ is not expected to be epistatic to factor $B$ as well.  \cite{cordell2002epistasis}

A mathematical definition of epistatic interaction as ``departure from neutrality" requires defining neutrality and measuring phenotype.

\begin{itemize}
	\item \textit{Phenotype} is often measured using the concept of fitness, particularly in many large-scale genetic interaction studies, since it is relatively easy to measure by population allele frequencies or growth rates of microbial cultures \cite{mani2008defining}. 
Different measures of fitness can be used in epistasis: i) exponential growth rate of mutant strain respect to wild type ; ii) the increase in population in one wild-type generation; and iii) the relative number of progeny (in one wild-type generation) \cite{mani2008defining}.
Based on this, four mathematically different definitions of interaction have been used (namely \textit{product}, \textit{additive}, \textit{log}, and \textit{minimum}) \cite{mani2008defining}, but even though some definitions yield identical results under some conditions, an alternative definition choice can lead to different consequences\cite{mani2008defining}.

	\item \textit{Neutrality} predicts the phenotype of an organism without interacting mutations. 
Genetic interaction studies have differed in their choice of neutrality functions, generally using either a multiplicative or a minimum mathematical function. Multiplicative function predicts fitness to be the product of the corresponding single-mutant fitness values. 
	The minimum definition is simply the minimum neutrality of the expected results form non-interacting mutations (e..g the fitness of the less-fit mutant). 
All the above fitness measurement yield the same set of genetic interactions under this neutrality definition. 
For example if each mutation disrupts a distinct pathway limiting cell growth in a way that one mutation is substantially more limiting than the other, the double mutant might is expected have same result as the most-limiting single mutant \cite{mani2008defining}.
\end{itemize}

It has been shown choice of definition can dramatically alter the resulting set of interactions \cite{mani2008defining}.
To evaluate this the authors applied all four definitions to two studies providing quantitative growth-rate measurements of cell populations. 
They show that: 
i) the \textit{additive} and \textit{log} definitions have different biases; 
ii) the \textit{product} and \textit{log} definitions are equivalent for deleterious mutations; 
iii) the \textit{product} definition can reveal functional relationships missed by the \textit{minimum} definition; and 
iv) interaction networks inferred based on the \textit{minimum} and \textit{product} definitions differ greatly. 
This leads to the question on which definition to use. 
By examining the deviation distribution of expected (double-mutant) phenotype from the observed phenotype they found that \textit{product} and \textit{log} definitions not only are the closest to the ideal, but also are practically equivalent when single mutants are deleterious.

The presence or absense of a trait are extreme aspects of ``perturbation in a complex system", but there are no reasons to expect all forms of epistasis to follow this pattern \cite{phillips2008epistasis}.
When applied to quantitative traits, epistasis also describes a situation in which the phenotype cannot be predicted by the sum of the phenotypes of its single-locus component \cite{carlborg2004epistasis}.
Many epistatic QTL interactions have been detected in model organisms leading to the conclusion that epistasis makes a large contribution to the genetic regulation of complex traits  \cite{carlborg2004epistasis}.

\subsection{Epistasis is ubiquitous}

One of the most common definition of epistasis is departure from additive effects.
Nevertheless, there is no reason to think that traits should be additive based on a purely biological perspective \cite{zuk2012mystery} since biology is riddled with non-linearity such as genetic networks exhibit binary states, ligand - receptors concentration having sigmoid-like response, concentration saturations of substrate - enzymes reactions, sharp transitions created by cooperative protein binding, the pathways constrained by rate-limiting inputs, etc. \cite{zuk2012mystery}. 
It has been asserted that epistatic effects are not isolated events, but ubiquitous \cite{tyler2009shadows} and probably inherent properties of biomolecular networks.
The thought epistasis in the classical sense may be ubiquitous has been partially confirmed from mutational studies \cite{phillips2008epistasis}.
Genetic studies of synthetic traits, which occur only when multiple loci or pathways are all disrupted, in model organisms have identified instances of interacting genes revealing that epistasis may be pervasive \cite{zuk2012mystery}. 
Researchers found \cite{phillips2008epistasis} that when looking for interactions induced by systematically over-expressing genes in Saccharomyces cerevisiae about $15\%$ of studied genes induced growth defects with most over-expression not matching the phenotypes of individual deletions.  

\subsection{Epistasis examples}

\paragraph{Non-human}
Several genotype-phenotype patterns are known to be caused by epistasis in animal and model organisms, classic examples include \cite{carlborg2004epistasis}:
coat colour in various animals, 
comb type in chickens, 
kernel colour in wheat, and
eye color in flies. 
Here we mention some examples:

\begin{itemize}

	\item Coat colour in mammals has been one of the most common examples. 
In pig, the dominant allele at the KIT locus confers white color coat and is dominant over all locus conferring darker color (melanocortin 1 receptor or MC1R). 
This can be determined in individuals with the recessive KIT genotype showing what was classically termed `dominant epistasis', 
yielding a non-Mendelian segregation ratio of 9:4:3 (instead of 9:3:3:1) \cite{carlborg2004epistasis, phillips2008epistasis}.

	\item Drosophila provides another classic example with eye color determination.
Drosophila eye pigmentation scarlet, brown, and white is determined by the synthesis of two drosopterins:  brown pigments (from tryptophan) and red pigments (from GTP) \cite{tyler2009shadows}.
A mutation that prevents production of the brown pigment results in a fly with red eyes and a mutation preventing red pigment results in a fly with brown eyes.
Flies with a mutation in the white gene, neither red nor brown pigment can be synthesized resulting in a fly with white eyes (regardless of the genotype at the brown or scarlet loci) \cite{tyler2009shadows}.

	\item Dozens of quantitative traits indicating strong epistasis in mouse and rat \cite{shao2008genetic} by analysing a panel of chromosome substitution strains where the effects attributed to the donor chromosomes exceeds by a median eightfold the expected effect of the donor genome.

	\item Genetic interaction have been studied in a systematic and large-scale manner in Saccharomyces cerevisiae \cite{jasnos2007epistatic,tong2001systematic}.
Analysis of quantitative traits loci (QTL) for transcripts levels in a two strain cross demonstrated epistatic interaction for $67\%$ of studied pairs (first the strongest QTL was found and then the strongest remaining QTL conditional on the first genotype was selected) \cite{brem2005genetic}. 

	\item In a study comparing three Drosophila inbred lines (Drosophila melanogaster Genetic Reference Panel) and a large outbred and intercross derived  population \cite{huang2012epistasis}, a set of candidate SNPs was selected by assesing allele frequency changes between the extremes of the distribution for each trait. 
The researchers found that the majority of these SNPs participated in at least one epistatic interaction \cite{huang2012epistasis}.
Using this information from epistatic interacting loci they were able to infer networks affecting quantitative traits. \cite{huang2012epistasis}.

\end{itemize}

\paragraph{Human}
Few instances of epistasis in common human disease have been discovered and well-replicated so far, despite considerable efforts \cite{zuk2012mystery}.
Although many instances of epistasis related to human disease have been published, with examples from type 2 diabetes\cite{wiltshire2006epistasis}, bipolar effective disorder\cite{jamra2007first}, coronary artery disease\cite{tsai2007renin}, and autism \cite{coutinho2007evidence}; some authors suspect these might be statistical features in the association studies because only a few have functional basis \cite{phillips2008epistasis}.
Perhaps the best examples of epistatic interactions in humans include:
\begin{itemize}
	\item Interactions involving at least one locus with a large effect such as HLA  \cite{zuk2012mystery}.
Two different interactions involving HLA alleles and ERAP have been discovered in GWAS from ankylosing spondylitis and psoriasis where the HLA alleles have odds ratio of 40.8 and 4.66 respectively \cite{evans2011interaction, genetic2010genome}.
In autoimmune disease multiple sclerosis researchers found evidence of genetic interactions between two histocompatibility loci known to be associated with the disease (HLA-DRB5*0101 in DR2a and HLA-DRB1*1501 in DR2b) \cite{gregersen2006functional}. 
In Type 1 diabetes HLA is assumed to act non-additively with all other risk alleles (HLA has have an effect of 5.5) \cite{barrett2009genome}.

	\item There was evidence of naturally occurring linkage disequilibrium which is suspected to be generated by strong epistasis \cite{phillips2008epistasis}.
In Hirschsprung's disease an interaction between RET and EDNRB was uncovered by a genome-wide linkage study (RET having a log-odds of 5.6). \cite{carrasquillo2002genome}

	\item The ACE gene (angiotensin I converting enzyme) has an epistatic interaction with AGTR1 gene (angiotensin II type 1 receptor ) gene, significantly increasing risk of myocardial infarction when the "D-allele" in ACE  is present in patients carrying a particular AGTR1 allele \cite{tiret1994synergistic}.

	\item Two different sets of interactions are assumed to be responsible for variation in triglyceride levels.
Notably, the interactions depend on the patient's sex: in females the interactions involves ApoB and ApoE;  and in males the interaction involves the ApoAI/CIII/AIV complex and low-density lipoprotein receptor \cite{nelson2001combinatorial}.

	\item Sickle-cell anemia is regarded as a Mendelian trait but is modified by epistatic interactions evidenced by the fact that patients homozygous for two polymorphisms near the $G\gamma$ locus have only mild clinical symptoms \cite{odenheimer1983heterogeneity}.

	\item Elevated blood serum cholesterol levels in humans is associated with an ApoE allele depending on the genotype at the LDLR (low density lipoprotein receptor) gene locus \cite{pedersen1989interaction}.

\end{itemize}

\subsection{Epistasis and evolution}

From an evolutionary perspective, some authors argue that the nonlinear nature of epistatic interactions between polymorphic loci is the genetic basis of canalization (the robustness or ability of a population to produce the same phenotype regardless of environmental variability) and speciation \cite{huang2012epistasis}.

It has also been pointed out that interactions have an important influence on evolutionary phenomena such as genetic divergence and affects the evolution of the structure of genetic systems \cite{phillips2008epistasis} based on studies and models showing that epistasis can have a limiting role on the possible paths that evolution can take \cite{miller2006direct}.

Theoretical grounds that date back to Fisher assert that when genes interact there is evolutionary pressure to promote their genetic linkage as a means of enhancing the co-inheritance of favourable allelic combinations \cite{fisher1958genetical}.
Under this assumption linkage can facilitate the maintenance of epistatic interactions and vice versa, thus explaining some molecular evolution complexity \cite{phillips2008epistasis}.
This has been supported by analysis of complex gene regulation patterns in localized genomic regions \cite{birney2007identification}.
For variety of organisms (such as yeast, Caenorhabditis, Drosophila, higher plants, and mammals) genes sharing expression patterns are more likely to be in proximity \cite{hurst2004evolutionary}.
This evidence shows that regional controls of chromatin structure and expression may give rise to gene clusters by promoting their coregulation \cite{petkov2005evidence}.

\subsection{Missing heritability}

At the dawn of the ``GWAS era" in 2002 it was hypothesised that there existed a large class of genetic traits for which GWAS would fail, namely purely epistatic models containing no additive or dominance variation at any of the susceptibility loci. 
Thus association case/control methods ``will have no power if the loci are examined individually" \cite{culverhouse2002perspective}.
Furthermore, it was mathematically shown that for such models maximizing the broad sense heritability (under some constraints) is equivalent maximizing gene interaction variance \cite{culverhouse2002perspective}.

In a seminal series of papers \cite{zuk2012mystery, zuk2014searching} further mathematical proof of the link between epistasis and heritability was provided.
The authors claim that missing heritability arises by an overestimation of the denominator that happens when epistasis is ignored \cite{zuk2012mystery}.
This overestimation, called ``phantom heritability", was shown to inflate the denominator over $60\%$ in Cohn's disease, thus accounting for up to $80\%$ of the missing heritability \cite{zuk2012mystery}.
Even though the prevailing view among geneticists is that interactions play at most a minor role in explaining missing heritability, their works shows that simple and plausible models can give rise to substantial phantom heritability \cite{zuk2012mystery}

In moderately heritable complex diseases for which single-locus GWAS analyses have not accounted for the predicted phenotypic variance these epistatic models provide one possible explanation so it is worth pursuing a hypothesis of interacting loci \cite{culverhouse2002perspective}.

\subsection{Detecting epistatic interactions}

Linkage disequilibria (LD) between close sites are the result of unrecombined chromosome blocks from common ancestry \cite{reich2001linkage}.
However LD between widely separated sites suggests epistatic selection forces are at work \cite{fisher1958genetical, koch2013long}.
In an analysis using Yorubian population (from Ibadan, Nigeria) of the HapMap dataset, patterns of LD were quantified and significance of overall disequilibrium per chromosome was evaluated of using randomization \cite{koch2013long}, showing an excess of associations in distant loci on all of the 22 autosomes. 
Although this is suggestive of epistasis, other hypotheses should not be ruled out:
i) population admixture has been proposed to explain unusual patterns of long range LD \cite{price2008long}
ii) recombination between distant chromosome blocks may not completely erase LD caused by drift even in a population at demographic equilibrium, 
iii) bottlenecks are particularly effective at generating long-range LD
iv) hitchhiking of linked sites with a positively-selected mutation can generate large haplotype blocks 
v) large inversion and other structural variation alter recombination patterns thus causing LD over unusually large regions \cite{bansal2007evidence}.

Under the assumption that long range LD can hint to epistasis due to physical protein interactions, the authors of LDGIdb \cite{wang2012ldgidb} created a catalogue of over $600,000$ pairs of SNPs showing strong long-range linkage disequilibrium, i.e. pairs of SNP pairs that were either located on different chromosomes or on different LD blocks and had $r^2 \ge 0.8$ \cite{wang2012ldgidb}.
However such a simple approach may be of little utility because of technical issues that must be taken into account when performing such association: i) commonly used measures of LD (such as $r^2$ and $D'$) are known to give rise to large linkage for variants with minor allele frequencies (MAF) near 0 \cite{koch2013long}; and ii) $r^2$ is not corrected for multiple testing.
A better alternative is to measure the probability that a large value of the disequilibrium $D$ is observed if there is no association, further refined by conditioning on the sampled allele frequencies at the two loci, which has the analytical advantage to asymptotically converge to a Fisher's exact test \cite{koch2013long}.

Another approach is to implicitly test over / under-representation of allele pairs in a given population, i.e. analysis of imbalanced allele pair frequencies \cite{ackermann2012systematic}.
The underlying theory is that such allele pairs are under Dobzhansky-Muller incompatibilities which establishes a fitness bias favouring of individuals that inherit over-represented allele combination \cite{ackermann2012systematic}.
Based on this, the authors in \cite{ackermann2012systematic} studied a population of 2,002 mice in family trios.
They performed a $\chi^2$ test correcting by confounding factors (such as allele frequencies, family structure and allelic drift) based on inspecting $3 \times 3$ contingency tables of all possible two-locus allele combinations.
They claim that their methodology can detect more interactions than using independent markers and as a result they were able to identify 168 LD block pairs with imbalanced alleles \cite{ackermann2012systematic}.

By exploiting the  intense selective pressures imposed by the process of inbred mice populations, it can be expected that clusters of functionally related genes are likely to be selected for coadapted allelic combinations in genes that influence fitness and survival.
This hypothesis would result in regions of linkage disequilibrium (LD) among inbred strain genomes that should occur more often than expected by chance \cite{petkov2005evidence}.
In a study using 60 inbred mouse strains \cite{petkov2005evidence}, the authors study LD using permutation tests and show that extreme patterns of LD give rise to scale-free networks architectures.
Further pathway analysis identifies biological functions underlying several of these networks, hinting that selective factors acting to generate LD networks during inbreeding are a reflect functional interaction \cite{petkov2005evidence}.

Although detection of epistasis is valuable, in this thesis we focus on the context of complex disease.
In the next sub-section we introduce methods combining GWAS and epistatic analysis to find epistatic loci affecting disease risk.

\subsection{Epistasis \& GWAS}

In recent years there have been a growing number of GWAS.
Most of them have used a single-locus analysis strategy, in which each variant is tested individually for association with a specific phenotype \cite{cordell2009detecting}.
Some researchers mentioned that it may be inadequate to describe relationships between genotype and complex disease by simply summing the modest effects from several contributing loci \cite{culverhouse2002perspective}.
Nevertheless, the extent to which epistasis is involved in complex traits is not known so we cannot assume that epistasis will be found for every trait in every population \cite{carlborg2004epistasis}.
However epistasis has been overlooked and should to be routinely explored in complex trait studies \cite{carlborg2004epistasis}.
This is particularly important for researchers of moderately heritable complex diseases for which locus-by-locus analyses have not accounted for the predicted genetic variance should pursue a hypothesis of epistatic loci \cite{culverhouse2002perspective} that owing to their interaction, might not be identified by using standard single-locus tests \cite{cordell2009detecting}.
It is also hoped that detecting such interactions will allow to elucidate biological pathways that underpin complex disease \cite{cordell2009detecting}.

Recent GWAS studies explored genome-level identification of epistatic interactions \cite{ackermann2012systematic}, even though methodological and sample size progress has been made, these could hardly identify a significant number of interactions. 
However failure to detect epistasis does not rule out it's presence \cite{zuk2012mystery}.
In theory a sufficient number of contributing purely epistatic interactions loci could account for all the variation in disease status for any prevalence \cite{culverhouse2002perspective}.
Nevertheless, when the genetic model of disease is purely epistatic (i.e. no additive or dominance at any of the susceptibility loci) then association methods analysing a single locus at a time cannot detect the loci \cite{culverhouse2002perspective}.
Furthermore there could be an \textit{n-way} purely epistatic model for which no joint analysis of two, three, or $n-1$ loci gives evidence.
This leads to the concern that even assessment of all ``$(n-1)$ way" interactions among candidate loci may not be sufficient for detection of the contributing loci \cite{culverhouse2002perspective}.

Another reason why complex human phenotypes fail to find evidence for epistatic interactions may simply be that analytic methods inherently exclude epistasis \cite{culverhouse2002perspective}.
For example individual interaction effects are expected to be much smaller than linear effects, and the sample size required to detect a variant scales inversely with the square of the effect size. 
The main obstacle is attributed to the exponentially large number of statistical hypotheses tested when comparing all markers against all other markers in a genome analysis \cite{ackermann2012systematic}.
As an example provided by Zuk et alii \cite{zuk2012mystery}, consider two variants with frequency $20\%$ and increasing risk by 1.3 fold, which is a large effect.
In such case, assuming $50\%$ power, a significance level of $5 \times 10^{-8}$  and equal number of cases and controls, the sample size required for single loci analysis would be $4,900$.
In comparison, the sample size required to detect pairwise interaction between those two variants using the same power and an appropriately corrected significance level is roughly $450,000$, so a researcher studying $100,000$ samples would discover all single acting loci but would find little evidence of epistatic interactions, which may be the reason why geneticists that have tested for pairwise epistasis have found few significant signals \cite{zuk2012mystery}.
It should be noted that even though GWAS involving over $500,000$ samples are not available at the moment, studies using sample sizes in this order are expected to become available within the next couple of years.

Existing approaches for identifying interactions in the context of GWAS can be grouped into five broad categories \cite{li2011detecting}:
\begin{enumerate}
	\item \textbf{Exhaustive search} methods extend classical single-locus GWAS statistics such as the Pearson's $\chi^2$ test or logistic regression.
For instance, using the definition of epistasis as ``departure from a linear model" \cite{cordell2009detecting}, in a logistic regression model the input for sample $s$ analysing loci $i$ and $j$ would include terms with each of the genotypes ($g_{s,i}$ and $g_{s,j}$), as well as an ``interaction term" ($g_{s,i} \cdot g_{s,j}$) \cite{cordell2002epistasis}. 

\begin{eqnarray*} \label{eq:gwasLogRegH1}
    P( d_s | g_{s,i},g_{s,j}) & = & \phi[ \theta_0 + \theta_1 g_{s,i} + \theta_2 g_{s,j} + \theta_3 (g_{s,i} g_{s,j}) \\
    & & ... + \theta_4 c_{s,1} + ... + \theta_m c_{s,N_{cov}} ] \\
\end{eqnarray*}

where $d_s$ is disease status, $\phi(\cdot)$ is the sigmoid function, $c_{s,1}, c_{s,2}, ... $ are covariates for sample $s$.
Logistic models involving interactions between more than two variants can be defined similarly, but require more parameters and extremely large samples are required to accurately fit them.

It should be noted that the number of tests necessary to evaluate all two-way, three-way and four-way interactions for 30-60 candidate loci, has a range similar to the number of tests suggested for a single GWAS, thus searching for n-way interactions among all the markers would be impracticable \cite{culverhouse2002perspective}.
%Other approaches include: 
%combinatorial partitioning method \cite{nelson2001combinatorial}, 
%restricted partitioning method \cite{culverhouse2004detecting}, 
%multifactor-dimensionality reduction \cite{ritchie2001multifactor}, 
%multivariate adaptive regression spline \cite{cook2004tree}, and 
%backward genotype-trait association (BGTA)\cite{zheng2006backward}. 
%Unfortunately even though many of these look promising, many of them have only been tested on small data sets \cite{zhang2007bayesian}.
Other approaches \cite{nelson2001combinatorial, culverhouse2004detecting, ritchie2001multifactor, cook2004tree, zheng2006backward}
although promising they have only been applied on small data sets \cite{zhang2007bayesian}.

Furthermore, methods based on brute-force searches such as (combinatorial partitioning method \cite{nelson2001combinatorial}, and multifactor-dimensionality reduction \cite{ritchie2001multifactor}) are impractical for large data sets \cite{zhang2007bayesian}.
Nevertheless it was shown \cite{li2011detecting} that it can be feasible to perform GWAS level analysis for only two interacting sites and that simple methods explicitly considering interactions can actually achieve reasonably high power with realistic sample sizes under different interaction models with some marginal effects, even after adjustments of multiple testing using the Bonferroni correction.
	
Exhaustive search methods exists for identifying epistatic variants affecting continuous phenotypes (or quantitative trait loci -QTL-).
In this case, matrix algorithm optimizations can significantly speed up computations.
For instance FastEpistasis applies an efficiently parallelized QR decomposition to derive least squares estimates of the interaction coefficient and its standard error \cite{schupbach2010fastepistasis}.
This allows it to handle all pairs from $500,000$ in a population of $5,000$ individuals in roughly one CPU year, which can be run in about a day on a 256 CPU cluster \cite{schupbach2010fastepistasis}.

	\item \textbf{Linkage disequilibrium} methods use patterns in disease population under two-locus disease models \cite{zhao2006test}.
	Association can be estimated assuming that deviation of the penetrance from independence at an individual locus creates linkage disequilibrium (LD) even if two loci are unlinked \cite{zhao2006test}.
In Zhao et alii \cite{zhao2006test} the authors, under the assumption that two disease-susceptibility loci are in Hardy-Weinberg equilibrium (HWE), show that in the presence of interaction the two loci will be in linkage disequilibrium in the disease population.
They develop a test statistic for the detection of deviations from LD by comparing the difference in the LD levels between two unlinked loci between cases and controls.
Assuming that the frequency of a haplotype is equal to the product of the frequencies of the component alleles of the haplotype, in the absence of interaction the proportion of individuals carrying a haplotype in the disease population is equal to the product of the proportions of individuals carrying the component alleles of the haplotype in the disease population.
Under the null hypothesis, this test statistic asymptotically converges to a central $\chi^2$ distribution.
Their power comparison simulations suggest that in general this LD-based test statistic has much smaller p-values than those of logistic regression analysis concluding that their test has higher power than alternatives such as logistic regression.
Nevertheless, their model does not account for cofactors, thus making it unsuitable in multi-ethnic GWAS where population stratification may confound disease risk.

	\item \textbf{Stochastic search} methods use sampling to infer whether a locus is an individual risk locus, epistatically affects disease risk, or has not effect (i.e. background locus).
A Bayesian approach for genome-wide case-control studies denoted `bayesian epistasis association mapping' (BEAM) \cite{zhang2007bayesian} is a representative example of this type of methods.
BEAM treats the disease-associated markers and their interactions via a bayesian partitioning model and computes the posterior probability using Markov chain Monte Carlo.
The method uses predictors in the form of genetic marker loci divided into three groups: 
i) markers not associated with disease, 
ii) markers individually contributing to disease risk, and 
iii) markers that interact with each other \cite{zhang2007bayesian}.
Membership of each marker in each of the three groups is defined by the prior distributions.
Given a prior distributions for regression coefficients values given by group membership, a posterior distribution can be generated using MCMC simulation \cite{cordell2009detecting}.
At the end, it uses a special statistic (B-Statistic) to infer significance from the samples in MCMC. 
Although it avoids explicitly computing all interactions the method could in theory find high-order interactions. 
Since BEAM was originally designed for genotypes markers, its power can be hampered by allele frequency discrepancies between unobserved disease loci and associated genotyped markers.
This is a common problem when using indirect markers and the authors show that in an extreme case when the MAF discrepancy was maximized all tested methods had little power to detect interaction associations.
In the original paper, the authors apply BEAM to a data set containing $116,204$ SNPs genotyped for $96$ affected individuals and $50$ controls for an association study of age-related macular degeneration (AMD).
Unfortunately BEAM did not find any significant interactions, most likely due to the small sample size.
Runtime and power are primarily determined by the number of MCMC rounds with a suggested number of MCMC iteration as the quadratic of the number of SNPs.
This is a main factor limiting the applicability \cite{li2011detecting}, so BEAM cannot easily be applied to large GWAS studies because computational limitations make it unsuitable to handle over $500,000$ markers with sample sizes of $5,000$ or more individuals, which are now commonly sequenced or genotyped \cite{cordell2009detecting}.

	\item \textbf{Conditional search} methods usually perform analyses in stages \cite{li2011detecting}.
A small subset of significant loci is identified in the first stage, typically using single locus association statistics.
Then this subset is mined using multi-locus association using an exhaustive method. 
A well known approach in this category is ``stepwise logistic regression" which works as follows:
i) all markers are individually tested for associations with disease using a logistic regression model;
ii) loci are ranked based on the results of single-locus tests;
iii) the top (usually $10\%$) are selected for epistatic association, 
iv) all two-way (or three-way) interactions are tested. 
Even this stepwise approach can become computationally intractable for high-order interactions \cite{zhang2007bayesian}.
Analysis of stepwise logistic regression approach to identify two-way and three-way interactions demonstrated that searching for interactions in genome-wide association mapping can be more fruitful than traditional approaches that exclusively focus on marginal effects \cite{zhang2007bayesian}.
As a counter argument for stepwise logistic regression, we should take into account that in the presence of epistasis the effect of one locus is altered or masked by another locus, thus power to detect the first locus is likely to be reduced and the joint effects will be hindered by their interaction \cite{cordell2002epistasis}. 
Methods based on conditional search can greatly reduce the computational burden by a couple of orders of magnitude, but with the risk of missing markers with small marginal effect \cite{li2011detecting}.

	\item \textbf{Machine learning} approaches can also be used to infer epistasis.
A popular approach uses Random Forests \cite{li2011detecting} or other regression trees partitioning approaches based on classification.
In this context, trees are constructed using rules based on the values of a predictor variable such as a SNP to differentiate observations such as case-control status \cite{cordell2009detecting}.
A popular rule selection mechanism is to use the variable that maximizes the reduction in Gini impurity \cite{kuhn1995application} at each node (intuitively, when child nodes have lower impurity from a split based on an attribute each child node will have purer classification).
Random forests are constructed by drawing samples with replacement from the original sample. 
A classification tree is created for each bootstrap sample, but only a random subset of the possible predictor variables is considered. 
This results in a `forest' of trees have been trained on a particular sample of observations. \cite{cordell2009detecting}.
Instead of trying to create a monolithic learner, this type of methods called ``ensemble systems" attempts to create many heterogeneous ``weak" (or simple) learners. 
The outcomes of these heterogeneous systems are combined to create an improved model \cite{li2011detecting}.

In Li et alii \cite{li2011detecting}, the authors create an extension of the AdaBoost algorithm where they incorporate an importance score based on Gini impurity to select candidate SNP  in a way that genotype frequencies from the two classes (case and control) are expected to be more different.
Decision trees are usually built with binary splits, but since genotype data takes three possible values $\{0, 1, 2\}$, they also extended their method to create a ternary split.
AdaBoost draws bootstrap samples to increase the power of a weak learner by weighting the individuals when bootstrapping. 
So when a weak learner misclassifies an individual, the weight of that individual is increased, as a consequence hard to classify individuals are more likely to be included in future bootstrap samples. 
The ensemble votes by weighting weak learner instances by training set accuracy.  \cite{li2011detecting}.
Using simulation, they claim that their method outperforms not only a similar ensemble approaches, but also several statistical methods, although they mention performance degradation when the risk allele frequency is low \cite{li2011detecting}.

\end{enumerate}

Although all these models have advantages under some assumptions, none of them seems to be a ``clear winner" over the rest \cite{cordell2009detecting}. 
All of these models suffer from the increase in number of tests that need to be performed, which raises two issues: 
i) multiple testing, and 
ii) computational feasibility.
So far, no method for epistatic GWAS has been widely adopted and there is need of different approaches to be explored. In Chapter \ref{ch:gwas} we propose an approach to combine co-evolutionary models and GWAS epistasis of pairs of putatively interacting loci.

\subsection{Conclusions \label{sec:epigwas}}

% GWAS & Epistasis
Genome wide association studies have traditionally focused on single variants or nearby groups of variants. 
An often cited reason for the lack of discovery of high impact risk factors in complex disease is that these models ignore interactions among loci \cite{cordell2009detecting} which have recently been pointed out as a potential solution for the ``missing heritability" problem \cite{zuk2012mystery, zuk2014searching}. 
With interactions being so ubiquitous in cell function, one may wonder why they have been so neglected by GWAS. 
There are several reasons: 
i) models using interactions are much more complex \cite{gao2010classification} and by definition non-linear, 
ii) information on which proteins interacts with which other proteins is incomplete \cite{venkatesan2009empirical}, 
iii) in the cases where there protein-protein interaction information is available, precise interacting sites are rarely know \cite{venkatesan2009empirical}, and
iv) protein interactions are not the only sources of epistatic loci, other types of interaction loci are less known and may be even harder to map.
Due to the lack of knowledge about interaction loci, we need to explore all possible loci combinations, thus the number of $N$ order interactions grows as $O(M^N)$ where $M$ is the number of variants \cite{de2013emerging}.
This requires exponentially more computational power than single locus models.
This also severely reduces statistical power, which translates into requiring larger cohort, thus increasing sample collection and sequencing costs \cite{de2013emerging}.

In Chapter \ref{ch:gwas} we develop a computationally tractable model for analysing putative interaction of pairs of variants from GWAS involving large case-control cohorts of complex disease. 
Our model is based on analysing cross-species multiple sequence alignments using a co-evolutionary model in order to obtain informative interaction prior probabilities that can be combined to perform GWAS analysis of pairs of non-synonymous variants that may interact.
