%---
\section{Coevolution}
%---

In a book published in 1859 entitled \textit{``On the origin of species by means of natural selection"} \cite{darwin1859origin}, Charles Darwin introduced the concept of co-evolution referring to the coordinated changes occurring in pairs of organisms.
In another of his books \textit{``On the various contrivances by which British and foreign orchids are fertilised by insects"}, first published in 1862  \cite{darwin1877various}; Darwin further explored this concept and providing more detailed examples.
By observing the relationship between the size of orchids' corolla and the length of the proboscis of pollinators, Darwin predicted the existence of a new species able to suck from a large spur  which was later confirmed \cite{de2013emerging}.

Co‐evolution originally referred to the coordinated changes occurring in pairs of organisms to improve or refine interactions.
This concept was later extended to pairs of proteins or more generically, any pair of biomolecules which can be within the same organism \cite{de2013emerging}.
The modern use of co-evolution methods in genetics is often attributed to Dobzhansky's \cite{dobzhansky1950genetics} and Elrich's \cite{ehrlich1964butterflies} seminal works that were published in 1950 and 1964 respectively.
In recent years, much effort has been dedicated to research of coordinated sequence changes in proteins (and genes) were co‐evolution could be an important and widespread catalyst of fitness optimization \cite{de2013emerging}.

Distinct allele combinations in co-evolving genes interact to confer different degrees of fitness. 
If this fitness difference is large, selection for alleles could maintain allelic association even between unlinked loci \cite{rohlfs2010detecting}, thus co-evolving genes are expected to maintain their interaction by pressures favouring compensatory mutations \cite{rohlfs2010detecting}.
Under this hypothesis, genetic loci may be invariable due to their functional or structural constraints but these constraints may change subject to mutations in their functional counterpart \cite{fares2006novel}.
In many cases, selective advantages for a specific allele pair could fixate the optimal allele pair in the population \cite{rohlfs2010detecting}.

\paragraph{Co-evolution examples}
In the absence of a clear positive control, identifying gene pairs that is certainly co-evolving are a difficult task \cite{rohlfs2010detecting}.
Here, some well known examples of co-evolution in humans are introduced:

\begin{itemize}

\item HLA ligand and killer-cell immunoglobulin-like receptor (KIR) are two genes located on different chromosomes forming a well established interacting immune-response pair.
Their allele frequencies are highly correlated in human populations as one expects under allele matching selection \cite{single2007global}.

\item A remarkable similarity in the phylogenetic trees of ligands (such as insulin and interleukins) and their corresponding receptors was observed.
This co‐evolution is proposed to be required for maintaining their specific interactions \cite{pazos2001similarity}.

\item Researchers found that ligands and their G-protein coupled receptors have co-evolved so that each subgroup of ligands has a matching subgroup of receptors \cite{goh2000co}.
%Ligand-receptor co-evolution can be detected based on N-terminal and C-terminal phosphoglycerate kinase (PGK) which are covalently linked and form an active site at their interface, therefore, they must are inferred to have co-evolved to preserve enzyme function \cite{goh2000co}.

\item In Hsp90 and GroEL heat-shock proteins, co-evolution was detected in ``almost all" functionally or structurally important site \cite{fares2006novel}.

\item GroESL is involved in the folding of a wide variety of other proteins with the folding activity mediated by the co-chaperonin GroES  \cite{ruiz2013coevolution}.
It was recently shown that different overlapping sets of amino acids co-evolve between GroEL and GroES \cite{ruiz2013coevolution}.

\item Gamete recognition genes ZP3 and ZP19 are highly polymorphic among humans and located on different chromosomes.
Putative interaction was observed between these genes was recently inferred \cite{rohlfs2010detecting}.

\item Helicobacter pylori is the main cause of gastric cancer. 
Host-pathogen interaction accounted for most of the difference in the severity of gastric lesions in the populations analysed. 
For instance African H. pylori ancestry was relatively benign in population of African ancestry but was deleterious in individuals with substantial Amerindian ancestry \cite{kodaman2014human}.
This is in an example of co-evolution modulating disease risk.

\end{itemize}

\subsection{Basic co-evolution inference models}

In this section we review the first methods aimed to uncover co-evolution.
These ``basic methods" serve not only to understand the historical perspective but also they are the basis of more advanced methodologies described in section \ref{sec:coevGlobal}.

\paragraph{Phylogenetic tree similarity}
%Co‐evolution of interacting species, such as symbionts-hosts, predators-prey, and parasites-hosts, is assumed to be manifested by similarities in the phylogenetic trees \cite{de2013emerging}.
Proteins and their interaction partners co-evolve so that divergent changes in one are complemented their interaction partner. 
These changes can be manifested by ``similar evolutionary trees" \cite{goh2000co}.
Thus phylogenetic similarity approaches can successfully be applied for protein-protein co‐evolution assumed to be caused by physical interactions.
These kind of methods have been shown to be capable of identifying interaction partners, such as ligand-receptor pairs \cite{de2013emerging}.

Similarly, evolutionary relationships within protein families can be mined to predict physical interaction specificities \cite{ramani2003exploiting}.
Duplicate genes (paralogs) can diverge in a way such that new binding specificities develop, thus the underlying hypothesis is that interacting proteins exhibit coordinated evolution and tend to have similar phylogenetic trees.
This was first demonstrated in a study of chemokines and their receptors showing phylogenetic tree similarities \cite{goh2000co}.
%Using similarity of phylogenetic trees as a proxy for the co-evolution of interacting proteins \cite{ramani2003exploiting}, a computational method based on matrix alignment can find an optimal alignment between protein family similarity matrices (conceptually equivalent to superimposing  phylogenetic trees from the two protein families) \cite{ramani2003exploiting}.
%One matrix is shuffled using stochastic simulated annealing-based to make the two matrices maximally agree by minimizing the root mean square difference.
%Interactions can be predicted by observing equivalent columns proteins heading in the two matrices.  \cite{ramani2003exploiting}

\paragraph{Correlated mutations}
Although some methods based on phylogenetic tree similarity exists, the majority of co-evolutionary methods focuses on analysis of multiple sequence alignment \cite{rohlfs2010detecting}.
Proteins have evolved to interact or function in specific molecular complexes and the specificity of these interactions is essential for their function. Consequently, residue contacts constrain the protein sequences to some extent \cite{pazos1997correlated}.
In other words, 
%sequences form interacting proteins react as a consequence of adaptation, thus 
it is reasonable to assume that evolution of sequence changes on one of the interacting proteins must be compensated by mutations in the other \cite{pazos1997correlated}.
It should be noted that this relationship between co-evolution and interaction is not symmetrical
%While interaction would involve coevolution, 
since co-evolution does not imply physical interaction \cite{fares2006novel}.
This is emphasized by the fact that co-evolution between clusters of sites not in contact has also been shown \cite{pritchard2000proteins}.

%Identification of genes showing signs of adaptive evolution can be used in determining functional regions in proteins \cite{fares2006novel}.
It has long been suggested that correlations in amino acid changes can be used to infer protein contact, thus helping predict tertiary protein structure \cite{fitch1970improved, morcos2011direct, burger2010disentangling, de2013emerging}.
A large number of genomes and protein sequences have become available in recent years enabling the analysis of co-evolution by means of statistical inference of covariation patterns based on multiple protein sequence alignments \cite{burger2010disentangling, burger2010disentangling}, which has been a fruitful technique for predicting contacting residues in the structure.
This interdependent changes in amino acids was formulated for the first time by the ``covarion model" \cite{fitch1970improved} and applied in multiple sequence alignments of a family of homologue proteins \cite{de2013emerging}.
Statistical methods to find correlated mutations between pairs of proteins can identify putative interaction sites in protein pairs \cite{de2013emerging}, but we should keep in mind that correlated mutations suggesting compensatory changes between residues can be due to several factors different than direct contact, such as physical proximity, catalytic action, binding sites, or even maintaining folding stability.

One of the first attempts of statistical inference of co-evolving loci pairs was performed by Gobel et. al in 1994.
In their seminal paper they point out that \textit{``maintenance of protein function and structure constrains the evolution of amino acid sequences... [sequence alignments] can be exploited to interpret correlated mutations observed in a sequence family as an indication of probable physical contact in three dimensions"} \cite{gobel1994correlated}. 
They  analysed correlations between different positions in a multiple sequence alignment and used such correlations to predict contact maps.
In their study of 11 protein families they compare their results with experimentally validated contact maps determined by crystallography, showing prediction accuracy up to $68\%$.

The promise of developing methods for predicting amino acid contacting pairs from sequence information alone was radically different from and more applicable than traditional docking methods \cite{pazos1997correlated}.
This lead to the development of methods for detecting correlated changes in multiple sequence alignments with the primary objective of using them to detect protein interfaces in interacting molecules \cite{pazos1997correlated}, thus facilitating protein structure prediction.
It was demonstrated that the correlated sequence information was enough to select the right inter-domain docking solution amongst many alternatives.

Correlation and mutual information (MI) have been used to assess co-evolution but they do not take the evolutionary interdependence between protein residues into account \cite{fares2006novel}.
Phylogenetic relationships can inflate these co-evolutionary measures, thus one of main limitations of these methods has been their inability to separate phylogenetic linkage from functional and structural co-evolution \cite{fares2006novel}.
Some methods partially correct these effects but while some studies \cite{gloor2005mutual} claim that these would require alignments of at least $125$ sequences, while other studies \cite{morcos2011direct} suggest that they may require in the order of $1,000$.

\paragraph{Phylogenetic correction}
Mutual information (MI) measures the reduction of uncertainty about one position given information about the other.
When used as a measurement for co-evolution, MI can be confounded by several factors such as structural and functional constraints, and the background sum of contributions from random noise and shared ancestry.
In an attempt to improve MI's signal to noise ratio by eliminating or minimizing the second factor, a model postulated by Dunn et alii \cite{dunn2008mutual} tries to factorize these terms in order to estimate a correction.
They propose that each amino acid position in the MSA has a propensity toward the background MI (related to its entropy and phylogenetic history) and estimate the joint background MI as the product of their propensities.
It follows that a joint background correction term can be approximated as product of the average background MI divided by the average overall MI of all positions in the MSA, which they call the average product correction (APC) \cite{dunn2008mutual}.
They show that APC is a metric than can accurately estimate MI in the absence of structural or functional relationships (i.e. the null model) \cite{dunn2008mutual}.
Finally, by assuming the null model to be normally distributed, a p-value can be inferred using a Z-score.

Another method, CAPS \cite{fares2006novel}, compares transition probability scores from the blocks substitution matrix (BLOSUM) between two sequences at the sites being analysed for interaction.
An alignment-specific BLOSUM matrix is applied depending on the average sequence identity.
Co-evolution between protein sites is estimated by the correlation in the pairwise variability with respect to the mean pairwise variability per site \cite{fares2006novel}.
A limitation of this method arises when sequences are too divergent, since an alignment including highly divergent sequence groups could show unrealistic level of pairwise identity (BLOSUM values are normalized by the time of divergence between sequences to reduce the impact of this).
Another problem common to many MSA-based co-evolutionary methods is that constant amino acid sites, which are very likely to be functionally important, cannot be tested for  \cite{fares2006novel}.

\paragraph{Evolutionary timespan}
What is the appropriate evolutionary time scale required in a multiple sequence alignment in order to perform a co-evolutionary analysis?
Co-evolution is often analysed over very large time frames based on the evolutionary analysis across different species \cite{qian2015recent}.
Nevertheless, genome-wide scans have identified several candidate loci that underlies local adaptations, which seems surprising given the short evolutionary time since the human divergence which is estimated have happened around $50,000$ to $100,000$ years ago when humans migrated out of Africa \cite{qian2015recent}.
In light of this, it may make sense to analyse co-evolution within human populations based on the propositions that multiple genes within a pathway or a functional sub-network may change in the same fitness direction at a same evolutionary rate to achieve a common phenotypic outcome \cite{qian2015recent}.
In a study using data from the 1000 Genomes project \cite{10002012integrated} form East Asians, Europeans, and Africans populations, researchers found that genes having signals of recent positive selection are significantly closer to each other within protein-protein interaction (PPI) networks \cite{qian2015recent}.
The approach was also able to identify known examples such as EGLN1 and EPAS1 (hypoxia-response pathway playing key roles in adaptation to high-altitude) as well as multiple genes in the NRG-ERBB4 (developmental) pathway \cite{qian2015recent}.
This shows that sequences from shorter time spans can also be mined for co-evolution.

\paragraph{MSA quality influences predictions}
Since many co-evolutionary methods rely so heavily on multiple sequence alignments, it should not be surprising to know that the quality of the input alignment may affect the results.
As one example, it is well known that structure-based alignment algorithms may be susceptible to shift error and other systematic errors, thus strong covariation signal can be caused by alignment errors leading to false positive predictions \cite{dickson2010identifying}.
The phylogeny of the sequences also affects performance, since methods work better on large protein families having a wide but homogeneously distributed degree of sequence similarity ranging from distant to similar sequences \cite{de2013emerging}.
In a recent study co-evolutionary methods applied to different alignments of the same protein family gave rise to distinct results, demonstrating that the measurement of co-evolution may greatly depend on the quality of the sequence alignment \cite{dickson2010identifying}.
Even when alignments for the same protein family contained comparable numbers of sequences the number of estimated co-varying positions differed significantly.
The authors of this analysis demonstrated that contact prediction can be improved by removing alignment errors due to several factors such as partial or otherwise erroneous sequences, the presence of paralogous sequences, and improper structure alignment.

\paragraph{Co-Evolution and protein structure}
Protein structure prediction from amino acid sequence is one of the ultimate goals in computational biology \cite{burger2010disentangling}, despite significant efforts the general problem of de novo three-dimensional structure prediction has remained one of the most challenging problems in the field \cite{marks2012protein}.
Unfortunately, \textit{de-novo} protein structure prediction does not scale with longer proteins since the conformational space grows exponentially with the protein length.
Inter-residue contact information can constrain the fold thus significantly reducing the search space.
Since covariation patterns can complement experimental structural biology thus helping to elucidate functional interactions, information of co-evolutionary couplings between residues are often used to compute protein three-dimensional structures from amino acid sequences \cite{marks2012protein}.
It has been observed that information about protein residue contacts, can be used to elucidate the fold of some proteins \cite{jones2012psicov}.
Researchers demonstrated that using co-evolutionary information from multiple sequence alignments greatly helps to deduce which amino acid pairs are close (or in contact) in the three-dimensional structure thus allowing the protein fold to be determined with a reasonable accuracy \cite{marks2012protein}.
It is not surprising that the vast majority of methods for finding protein co-evolution are designed with the specific aim of generating results useful in the context of protein folding.

\paragraph{Protein design}
It has recently been proposed to use co-evolutionary theory in computational methods for protein design.
Significant similarities were found between the amino acid covariation in natural protein sequences and sequences structures optimized by computational protein design methods \cite{ollikainen2013computational}.
Because evolutionary selective pressures on function and structure shaped the sequences to be close to optimal for their structures, natural protein sequences provide an excellent source for computational protein design methods.
%Similarly, computational protein design predicts energetically optimal sequences based on protein structure, so it is expected that highly co-varying amino acids pairs in both designed and natural sequences have co-varied to maintain optimal protein structure.
This study using computational protein design to quantify protein structure constraints from amino acid covariation for 40 diverse protein domains, shows that structural constraints imposed by covariation play a dominant role in protein architecture.
Computational protein design methods could make use of knowledge form natural co-evolution effects \cite{ollikainen2013computational}.

\subsection{Global co-evolution models \label{sec:coevGlobal}}

Imagine a protein sequence of length $L = a_1, a_2, ... , a_n$, amino acid $a_i$ is coupled directly with $a_j$, and $a_j$ to $a_k$, then $a_i$ and $a_k$ will show correlation despite not being directly coupled \cite{weigt2009identification}.
This is an important problem when inferring co-evolution as indirect coupling can make it difficult to recognize the directly co-evolving loci.
%occurring when more than two positions show coordinated substitution patterns.
%Apparent co‐variation between two positions is the consequence of the evolutionary interdependence and these indirect couplings can make it difficult to recognize the directly interdependent positions.

As opposed to models using the independence assumption, a `global' model treats correlated pairs of residues as dependent on each other thereby minimizing effects of transitivity  \cite{marks2012protein}.
Since direct couplings are more reliable predictions of physical interactions, approaches that can distinguish direct from indirect couplings have been an intensive area of study \cite{de2013emerging}.
Global approaches are designed to reach high scores only for amino acid pairs that are likely to be causative of the observed correlations  \cite{marks2012protein}.
In this section we introduce these methods.


\paragraph{Glass spin systems}
Global interaction models are well understood in statistical physics.
A typical example are long-range order observed in spin systems, where the spins only have short-range direct interactions \cite{binney1992theory}.
One of the first global models for co-evolution was proposed by Lapedes \cite{lapedes2012using}, who used a Monte Carlo algorithm to infer the simplest probabilistic distribution able to account for the whole network of co‐variations \cite{de2013emerging}.
He presented a sequence-based probabilistic theory addressing co-operative effects in interacting positions in proteins assuming that a sequence of length $L$ is a global state of an \textit{L-site} spin system of twenty states (for twenty amino acids).
Then he solved the global statistical formalism based on maximizing entropy under constraints which are known to lead to Boltzmann statistics \cite{marks2012protein}.
Finally the conditional mutual information is calculated using this Boltzmann model which leads to the degree of covariation between residues at two positions factoring out contributions by interaction with the rest of the residues \cite{marks2012protein}.
The amount sequence data is a limiting factor when performing inference of Boltzmann distribution parameters, thus it is usually infeasible to use more than first order distributions \cite{lapedes2012using}.
Another limitation is the phylogenetic relatedness of these sequences, which is not addressed in this algorithm and has the potential to decrease accuracy \cite{lapedes2012using}.

\paragraph{Direct coupling analysis}
A similar approach called  direct-coupling analysis (DCA) was also based on spin-glass physics \cite{weigt2009identification}.
In their implementation a generalized message-passing technique is used to massively parallelize the algorithm implementation.
As in in the work of Lapedes \cite{lapedes2012using} an application of the maximum entropy principle yields the Boltzmann distribution which is used to estimate the second order interaction model.
In principle higher correlations of three or more positions can be included, however dataset size (i.e. number of sequences in the MSA) does not allow for inference beyond two-residue model parameters. 
Determining model parameters, which is the most computationally expensive task is achieved by using a two-step procedure: 
i) given a candidate set of model parameters, single and two residue distributions are estimated; 
ii) the summation over all possible protein sequences would require $O(|\Sigma_{AA}|^{N-2} N^2)$ steps (where $\Sigma_{AA}$ is the amino acid alphabet and $|\Sigma_{AA}|$ is the alphabet size), so an approximation is performed using MCMC sampling. 
This last step is the most expensive step and is expected to be very slow for 21-state variables.
The message-passing approach implemented using an efficient heuristic, reduces the computational complexity to $O(|\Sigma_{AA}|^2 N^4)$.
Once all probability distributions are estimated, gradient descent is used to adjust the coupling strengths maximizing the joint probability of the data.
Since the model is convex, it is guaranteed to converge to a single global maximum.
Finally, a quantity called direct information (DI) measures the part of the mutual information of a position pair induced by the direct coupling (intuitively similar to mutual information in a two-variable model).
Even after all optimizations and parallelizations, the method could not be applied to more than $60$ positions in the protein alignment simultaneously.
The authors apply the method to a dataset consisting of over $2,500$ bacterial genes from a two-component signal transduction system,. Their global inference robustly identified residue pairs proximal in space between sensor kinase (SK) and response regulator (RR) proteins as well as homo-interactions in RR proteins \cite{weigt2009identification}.
In their test dataset, the top $10$ candidate interactions identified were shown to be true contacts, furthermore these predictions were then used to calculate an interacting protein complex quite accurately (3 \AA\  RMSD) \cite{weigt2009identification}.

\paragraph{Mean field approximation}
DCA has been shown to yield a large number of correctly predicted contacts based on its ability to disentangle direct and indirect correlations; unfortunately the method is computationally expensive \cite{weigt2009identification}.
A method published by Morcos et alii \cite{morcos2011direct} proposes a ``mean field" approximation to DCA \cite{weigt2009identification}.
They first attempt to mitigate phylogenetic tree biases using a simple sampling correction based on re-weighting  sequences with more than $80\%$ similarity.
In a nutshell, the approximation method also tries to disentangle direct and indirect couplings by inferring a global statistical and least-constrained model which, as discussed before, is achieved using a maximum-entropy principle leading to a Boltzmann distribution of couplings.
The partition function ($Z$) is then approximated by keeping only the linear order term in a Taylor series expansion, thus obtaining the mean-field equations.
This approach is based on small-coupling expansion, thus a Taylor expansion around zero, a technique introduced in disordered Ising spinglass models with binary variables.
A well known result is that the first derivative of the Gibbs potential, the Legendre transform of the free energy $F = - ln(Z)$, equals the average of the coupling term in the Hamiltonian.
This simplifies this average calculation since the joint distribution of all variables becomes factorized over the single sites \cite{morcos2011direct}.
This mfDCA algorithm speeds up the original DCA implementation by $10^3$ to $10^4$ times \cite{morcos2011direct}, and can run on alignments up to $500$ amino acids per row which is an order of magnitude larger the previous version of DCA based on message passing \cite{morcos2011direct, weigt2009identification}.

\paragraph{PSI-COV}
Like other methods, PSI-COV \cite{jones2012psicov} starts from a multiple sequence alignment.
A covariance matrix is calculated by counting how often a given pair of amino acids occurs in a particular pair of positions, summing over all sequences in the MSA.
Since this matrix contains the raw data capturing all residue pair relationships, one can then compute a measure of causative correlations in the global statistical approaches by taking the inverse of the covariance matrix \cite{jones2012psicov, marks2012protein}.
Assuming that this covariance matrix can indeed be inverted, the inverse matrix relates to the degree of direct coupling, a well known fact in statistical theory under the assumption of continuous Gaussian multivariate distributions \cite{marks2012protein}.
Elements significantly different from zero (off-diagonal) indicate pairs of sites which have strong direct coupling and are thus likely to be in direct physical contact \cite{jones2012psicov}.
Unfortunately, the empirical covariance matrices are actually almost always singular simply because it is unlikely that every amino acid is observed at every site.
One of the most powerful techniques to overcome this problem is sparse inverse covariance estimation under Lasso constraints.
The authors claim that the non-zero terms tend to more accurately relate to correct correlations in the true inverse covariance matrix \cite{jones2012psicov}.

\paragraph{Multidimensional mutual information}
In a recent study a simple extension of mutual information was proposed by considering ``additional information channels" corresponding to indirect
amino acid dependencies \cite{clark2014multidimensional}.
This is achieved by defining the information $I(X_1 ; X_3 ; X_2)$ representing an `interaction information' for a channel with two inputs $X_1$ and $X_3$ and a single output $X_2$.
The effect of the indirect input ($X_3$) on the transmission between $X_1$ and $X_2$ can then be marginalized simply by summing mutual information for each possible value $X_3$ weighted by the probability of occurrence \cite{clark2014multidimensional}.
Similarly a four variable model extension can be defined, in which case the marginalization would be done over two variables ($X_3$ and $X_4$).
The authors test and compare their results using a set of $9$ MSAs consisting of less than $400$ sequences each, showing that their simple extension is comparable to other maximum entropy statistical models \cite{clark2014multidimensional}.
Even thought the method is simple, the marginalization sums impose a heavy computational burden requiring long execution times and large memory footprints making the method impractical for sequences longer than 200 residues \cite{clark2014multidimensional}.

\paragraph{Bayesian network model}
Another attempt to disentangle direct from indirect statistical dependencies between residues assumes that the sequences in a MSA are drawn from unknown joint probability distribution \cite{burger2010disentangling}. 
The model considers pairwise conditional dependencies and factorizes the joint probability by a single other position which the residue depends on, using the conditional probabilities as nuisance parameters that are integrated out when calculating the likelihood of the alignment. 
Most notably, the model does not consider only the best way of choosing the dependent position, but rather sums over all possible ways in which dependencies could be chosen.
This sum over all spanning trees is a generalization of Kirchhoff's matrix-tree theorem and can be efficiently computed form the Laplacian of the dependency matrix.

\subsection{Algorithm limitations}

%Residue co‐evolution was originally detected using correlated amino acid changes in pairs of positions represented by two columns of the MSA.
%Under the assumption of interdependent amino acid frequencies or similar patterns of amino acid substitutions it can be assessed by a linear correlation, a method that shows a small but significant capability to recover pairs of positions in physical contact \cite{de2013emerging}.

%MI
Mutual information was one of the first proposed methods used to detect co‐varying positions. 
As opposed to correlation-based methods, mutual information considers the distribution of each amino acid in the different sequences for a position quantifying whether presence of an amino acid one position can be used to predict presence of an amino acid in the other position.
Mutual information does not take into account which amino acids are present, therefore different amino acids are treated just as symbols \cite{de2013emerging}.
MI is an attractive and simple metric because it explicitly measures the dependence of one position on another, but it is limited by factors such as: 
i) positions with higher entropy (variability), tend to have higher MI than positions of lower entropy even though the latter are more constrained and would seem more likely to be co-evolving \cite{dunn2008mutual}; and 
ii) MI arises when alignments do not contain enough sequences to reduce the noise to signal ratio, it was shown that alignments should contain at least 125 sequences to significantly reduce this effect \cite{martin2005using}.

% PHYLO
The influence of the background phylogenetic relationship between sequences in the MSA confounds results and some efforts have tried to address this by removing certain problematic clades from the MSA.
For instance, it has been shown that the effect may be limited to some degree by excluding highly similar sequences (from closely related species) from the alignment \cite{wollenberg2000separation}.
Continuous-time Markov process model for sequence co‐evolution can model this explicitly and some approaches have been implemented for small-scale studies of co‐evolution in small protein families, but computational limitations have hindered their usage in large-scale studies \cite{de2013emerging}.
Other confounding effect is an uneven representation of protein sequence members (e.g. having several small subgroups and one large subgroup) which leads to statistical noise \cite{marks2012protein}.

%Indirect correlations arise because if $A$ correlates with and $B$ are in contact with each other and $B$ and $C$ are in contact as well, there is an observed indirect correlation between $A$ and $C$ \cite{marks2012protein}.
Since amino acids often contact more than one amino acids, transitive effects tend to form a network.
Thus pairs of residues analysed using a simple statistical model (such as correlation or mutual information) may not necessarily be close in space or functionally constrained \cite{marks2012protein}.
Algorithms to overcome this limitation exists, but they are based in global probabilistic models which require parameter estimation of complex distributions, such as the Bolzmann distribution, as well as marginalizing over all indirect variables.
This makes global models computational prohibitively for all but very small datasets and impossible to apply to genome wide scale analysis.

Usually co-evolutionary methods are tested with high quality MSAs containing large number of sequences varying from $5L$ to $25L$ (where $L$ is  sequence length).
Such large number of homogeneous sequences are rarely available and when they are, they usually correspond to well studied proteins and might already have a crystallized structure, thus analysis of amino acids in contact are not needed to infer the 3-D structure.
Often, investigators study less well-characterized proteins having MSA of less than $L$ sequences, and low alignment quality due to the presence of many gaps, in which case, existing methods are of limited value \cite{clark2014multidimensional}.

Finally it should be mentioned that results from different models usually do not agree, even for complex global models.
In a recent study, a comparison of several methods shows that while all methods detected similar numbers of co-varying pairs 
%(when taking into account residues separated by $\le 8$ \AA\ in reference X-ray structures)
, there is less than $65\%$ overlap between the top scoring prediction from methods based on different principles \cite{clark2014multidimensional}.
